### PR TITLE
Upgradable

### PR DESCRIPTION
**Pull request by simonmorley**
This PR does the following:

- updates the hardhat config removing the mainnet fork which ain't necessary
- updates hardhat config for upgradability
- adds node_modules to gitignore
- adds upgradability to the contracts
- adds tests for this too....

### CODE REVIEW

File: contracts/test_utils/SEUROV2.sol [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+
+contract SEuroV2 is Initializable, ERC20Upgradeable, UUPSUpgradeable, OwnableUpgradeable, AccessControlUpgradeable {
+  bytes32 public constant MINTER_ROLE = keccak256("MINTER_ROLE");
+  bytes32 public constant BURNER_ROLE = keccak256("BURNER_ROLE");
+
+  /// @custom:oz-upgrades-unsafe-allow constructor
+  constructor() initializer {}
```
Comment (therealdannzor) [inline contracts/test_utils/SEUROV2.sol:15]: no more grant roles in initialiser?

File: hardhat.config.js [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+    hardhat: {},
+    goerli: {
+      url: `https://goerli.infura.io/v3/${INFURA_API_KEY}`,
+      accounts: [`${GOERLI_PRIVATE_KEY}`],
+      // gasMultiplier: 2,
+    },
+    mainnet: {
+      url: `https://mainnet.infura.io/v3/${INFURA_API_KEY}`,
+      // accounts: [`${MAINNET_PRIVATE_KEY}`],
+      // gasMultiplier: 2,
```
Comment (therealdannzor) [inline hardhat.config.js]: keep comment if we would need to uncomment as some point, if it's redundant let's remove it

**Review by therealdannzor — COMMENTED** [review]
nice

File: contracts/test_utils/SEUROV2.sol (hunk 1 above)
Comment (simonmorley) [reply contracts/test_utils/SEUROV2.sol:15]: you're looking at v2 which is in test_utils

File: hardhat.config.js (hunk 2 above)
Comment (simonmorley) [reply hardhat.config.js]: Saving for longevity otherwise it'll be squashed

File: hardhat.config.js [hunk 3]
```diff
@@ -2,8 +2,21 @@ require("@nomiclabs/hardhat-waffle");
...
 
 module.exports = {
-  solidity: '0.8.15'
+  solidity: '0.8.15',
+  networks: {
+    hardhat: {},
+    goerli: {
+      url: `https://goerli.infura.io/v3/${INFURA_API_KEY}`,
+      accounts: [`${GOERLI_PRIVATE_KEY}`],
+      // gasMultiplier: 2,
```
Comment (ewansheldon) [inline hardhat.config.js]: do we need these comments in here still?

File: scripts/01_deploy.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+const { hre, ethers, upgrades } = require("hardhat");
+
+async function main() {
+  const SEuroContract = await ethers.getContractFactory('SEuro');
+  SEuro = await upgrades.deployProxy(SEuroContract, ['sEURO', 'SEURO'], {kind: 'uups'});
```
Comment (ewansheldon) [inline scripts/01_deploy.js]: declare this properly, would you

File: scripts/02_upgrade.js [hunk 5]
```diff
@@ -0,0 +1,18 @@
+const { hre, ethers, upgrades } = require("hardhat");
```
Comment (ewansheldon) [inline scripts/02_upgrade.js]: hre is unused

File: scripts/01_deploy.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+const { hre, ethers, upgrades } = require("hardhat");
```
Comment (ewansheldon) [inline scripts/01_deploy.js]: hre unused

**Review by ewansheldon — COMMENTED** [review]
seems sane enough. nice tests

some nitpicking if you're interested

**Review by ewansheldon — APPROVED** [review]
👍🏻

**Pull request merged**